### PR TITLE
[iov-bns] Add query username support

### PR DESCRIPTION
**Pull request by webmaster128**
This is supposed to cover points 1 and 2 from https://github.com/iov-one/iov-core/issues/495

### CODE REVIEW

File: packages/iov-bns/src/bnsconnection.ts [hunk 1]
```diff
@@ -445,15 +477,23 @@ export interface QueryResponse {
   readonly results: ReadonlyArray<Result>;
 }
 
-function parseMap<T extends {}>(decoder: Decoder<T>, sliceKey: number): (res: Result) => T & Keyed {
-  const mapper = (res: Result): T & Keyed => {
+function createParser<T extends {}>(decoder: Decoder<T>, keyPrefix: string): (res: Result) => T & Keyed {
```
Comment (ethanfrey) [inline packages/iov-bns/src/bnsconnection.ts]: Yes, this is a better, cleaner implementation than the other

File: packages/iov-bns/src/decode.spec.ts [hunk 2]
```diff
@@ -18,9 +18,34 @@ import {
 } from "./testdata";
 import { decodePrivkey, decodePubkey } from "./types";
 
-const { toUtf8 } = Encoding;
+const { fromHex, toUtf8 } = Encoding;
 
 describe("Decode", () => {
+  it("decodes username NFT", () => {
```
Comment (ethanfrey) [inline packages/iov-bns/src/decode.spec.ts]: :+1:

File: packages/iov-bns/src/decode.ts [hunk 3]
```diff
@@ -23,11 +23,22 @@ import {
...
 import * as codecImpl from "./generated/codecimpl";
-import { asNumber, decodeFullSig, ensure } from "./types";
+import { asNumber, BnsAddressBytes, BnsUsernameNft, decodeFullSig, ensure } from "./types";
 import { encodeBnsAddress, isHashIdentifier } from "./util";
 
 const { fromUtf8 } = Encoding;
 
+export function decodeUsernameNft(nft: codecImpl.username.IUsernameToken): BnsUsernameNft {
+  return {
+    id: fromUtf8(nft.base!.id!),
```
Comment (ethanfrey) [inline packages/iov-bns/src/decode.ts]: What I often do in these cases, where the data comes from outside, is to use `ensure`, which will throw a more useful error than just `.id!`. (or at least should throw a better error)

Comment (ethanfrey) [reply packages/iov-bns/src/decode.ts]: ex: `const sig = ensure(sigs[0], "first signature");`

So error is "first signature is missing", not "cannot access id of undefined"

File: packages/iov-bns/src/types.ts [hunk 4]
```diff
@@ -1,12 +1,59 @@
...
 
-import { Algorithm, PublicKeyBundle, PublicKeyBytes, SignatureBytes } from "@iov/base-types";
-import { FullSignature, Nonce } from "@iov/bcp-types";
+import { Algorithm, ChainId, PublicKeyBundle, PublicKeyBytes, SignatureBytes } from "@iov/base-types";
+import { Address, ChainAddressPair, FullSignature, Nonce } from "@iov/bcp-types";
 import { Int53 } from "@iov/encoding";
 
 import * as codecImpl from "./generated/codecimpl";
 
+export interface BnsUsernamesByUsernameQuery {
```
Comment (ethanfrey) [inline packages/iov-bns/src/types.ts]: nice set of queries

File: packages/iov-bns/src/bnsconnection.ts [hunk 5]
```diff
@@ -416,6 +428,26 @@ export class BnsConnection implements BcpAtomicSwapConnection {
     );
   }
 
+  public async getUsernames(query: BnsUsernamesQuery): Promise<ReadonlyArray<BnsUsernameNft>> {
```
Comment (ethanfrey) [inline packages/iov-bns/src/bnsconnection.ts]: very nice function, i like this.
good use of the paths and the arguments.

you are entering weave mastery :)

Comment (webmaster128) [reply packages/iov-bns/src/bnsconnection.ts]: :)

File: packages/iov-bns/src/decode.ts (hunk 3 above)
Comment (webmaster128) [reply packages/iov-bns/src/decode.ts]: Thanks, right. Let's do what we always do. Will update for both decodeBlockchainNft/decodeUsernameNft in the other PR

**Comment by webmaster128** [comment]
Closing in favour of #550

**Pull request closed without merging**